### DISCUSSION TITLE
[Mobile] How can I add new message on last by manually

### DISCUSSION THREAD

**Discussion by ghost** [singerdmx/flutter-quill#616] [Mobile] [answered]
Dear Developers,

Thank you very much for your support and I am sorry I don't know how to add a new message at end of the data with attributes manually.

Could you please kindly tell me, how can I add a new message by inserting it at end of the document with the attribute manually?

Thank you very much

**Comment by MagTuxGit** [comment]
`delta.insert('a new message', {'bold': true});`

**Comment by ghost** [comment]
Thank you very much @MagTuxGit .
But when I use `quillController.document.toDelta().insert('(edited)', {'color' : '#888888'});`, it doesn't add that message.
Could you please kindly help me 🙏 
Thank you very much

**Comment by ghost** [comment]
I would like to show (edited) at end of message.

<img width="506" alt="Screen Shot 2022-01-25 at 6 48 38 PM" src="https://user-images.githubusercontent.com/32185240/150975722-d8a96ade-a01e-4d9b-bc69-9b6634036cd3.png">

<img width="398" alt="Screen Shot 2022-01-25 at 6 50 55 PM" src="https://user-images.githubusercontent.com/32185240/150976033-1b52d650-980a-4c8e-bf4a-4a913d6f43cd.png">

**Comment by ghost** [comment]
When I use the following code, it can show (edited) at end of message but last message ending with header 1, "(edited)" is also showing header attribute

```
      quillController!.document.insert(quillController!.document.length - 1,
          ' ${AppLocalizations.of(context)!.message_edited}');
      quillController!.document.format(quillController!.document.length - (edited.length),
          edited.length, ColorAttribute('#888888'));
      quillController!.document.format(
          quillController!.document.length - (edited.length), edited.length, SmallAttribute());
```

before adding "(edited)", log output:

```
[{"insert":"Test","attributes":{"background":"#004d40"}},{"insert":" "},{"insert":"22","attributes":{"code":true}},{"insert":"\n","attributes":{"header":1}}]
```

After adding "(edited)", log output

```
[{"insert":"Test","attributes":{"background":"#004d40"}},{"in
sert":" "},{"insert":"22 ","attributes":{"code":true}},{"insert":"(edited)","attributes":{"code":true,"color":"#888888","small":true}},{"insert":"\n","attributes":{"header":1}}]
```

<img width="426" alt="Screen Shot 2022-01-25 at 7 31 59 PM" src="https://user-images.githubusercontent.com/32185240/150981709-9e3a00a1-5001-4f49-84f5-b46b93a15c0d.png">

**Comment by MagTuxGit** [comment, marked as answer]
@tzthantzin 
Actually I don't understand what the result are you expecting. It would be better to specify what do you want to get.

**Reply by ghost** [reply]
Hi @MagTuxGit, 
Thank you very much for your support.
 
I would like to add "(edited)" at the end of message, but if the last message has the attribute of header or background, when I added "(edited)" this edited will also added header or background.

I would like to expect, "(edited)" message do not want to have any attribute.
Could you please kindly help me?

**Reply by MagTuxGit** [reply]
All block attributes apply to the whole line. You can't make (edited) not to be header. But you can replace header with bold + size, which are inline attributes.

**Reply by MagTuxGit** [reply]
Also you can try to add (edited) outside of QuillEditor

**Reply by ghost** [reply]
Thank you very much for your kindly support. @MagTuxGit

**Reply by ghost** [reply]
Hi Mr. @MagTuxGit 

I would like to one question, I don't know, how to remove last character manually. Could you please kindly help me pls

**Reply by MagTuxGit** [reply]
> Hi Mr. @MagTuxGit
> 
> I would like to one question, I don't know, how to remove last character manually. Could you please kindly help me pls

Something like this:
```
Delta newDelta = oldDelta.compose(Delta()..retain(length - 1)..delete(1));
```

or

```
Delta newDelta = document.delete(length-1, 1);
```

**Reply by ghost** [reply]
Thank you very much Mr.@MagTuxGit,

> Delta newDelta = oldDelta.compose(Delta()..retain(length - 1)..delete(1));

It is working well on text, but it is not ok to delete emoji '\ud83d', could you please kindly help me 🙏 

I would like to do as backspace.

I am using emoji_picker_flutter and add emoji to quillController 

```
messageListModel.insertQuillController!.document.insert(
        messageListModel
                .insertQuillController!.plainTextEditingValue.text.length -
            1,
        emoji.emoji);
```

**Reply by MagTuxGit** [reply]
> It is working well on text, but it is not ok to delete emoji '\ud83d', could you please kindly help me 🙏

```
Delta newDelta = oldDelta.compose(Delta()..retain(length - lastCharLength)..delete(lastCharLength));
```

**Reply by ghost** [reply]
Thank you very much Mr @MagTuxGit,

Firstly I am really sorry asking many questions.
How can I get lastCharLength?

```
Test 😄 👍 😃 💯 [Cursor]
When I would like to delete 💯  , How can I get 💯  charLength.
```

```
Test 😄 👍 [Cursor]😃 💯 
When I would like to delete 👍 , How can I get 👍  charLength.
```

```
Test 😄 👍 🔢  TEST [Cursor]😃 💯 
In this case, I would like to delete only "T" so how can I get "T" charLength.
```

```
Test 😄 👍 🔢  TEST 😃 
[Cursor]💯 
In this case, I would like to delete only "😃" so how can I get "😃" charLength.
```

My requirement is totally same with OS keyboard backpress case.

Could you please kindly help me 🙏

**Reply by MagTuxGit** [reply]
I believe there is another more correct way to resolve your issue, but I'll try to answer your question:

``` dart
    bool isSecondUtf16Surrogate(int value) {
      return value & 0xFC00 == 0xDC00;
    }
    int charLength = 1;
    if (isSecondUtf16Surrogate(stringValue.codeUnitAt(cursorPosition-1))) {
      charLength = 2;
    }
```
I didn't check this code so it may not work for you.

**Reply by ghost** [reply]
Thank you very much @MagTuxGit 

I think it is better to support the delete function in the library. If I would like to delete the previous char from the current cursor position, just call library.delete method and it will delete manually.

**Reply by MagTuxGit** [reply]
I don't understand what are you trying to achieve. Why do you want to delete manually?

**Reply by ghost** [reply]
Yes @MagTuxGit, I would like to do backspace remove character manually, because I using with emoji picker, in emojji picker when backspace, I would like to delete character from current focused cursor position at quill editor.

**Reply by ghost** [reply]
Thank you very much @MagTuxGit 
 ```
bool isSecondUtf16Surrogate(int value) {
      return value & 0xFC00 == 0xDC00;
    }
    int charLength = 1;
    if (isSecondUtf16Surrogate(stringValue.codeUnitAt(cursorPosition-1))) {
      charLength = 2;
    }
```
This is code is working well.

Thank you very much for many supporting to me 🙏
